### DISCUSSION TITLE
statusCode to reflect what is available in the openAPI

### DISCUSSION THREAD

**Discussion by savage-alex** [apideck-libraries/portman#238] [Ideas]
I am trying to create a generic catchall contract test starter.

One of the things I noticed was that when using statusCode

```
{
                "openApiOperation": "*::/*",
                "statusCode": {
                    "enabled": true
                }
            },
```
The test created in postman always requires 200 to be returned even if that is not what is in the OpenAPI. 
I understand statusCode takes in a code parameter but would it be beneficial if that is not present to allow any of the codes from that operations openAPI to be present? 

Love to hear your thoughts?

**Comment by thim81** [comment]
Hi @savage-alex,

The Portman contract tests are organised around the following concepts:
-  contract test = Happy path, verify the succes response which uses by default the 200 request OpenAPI specification to generate the tests: statusCode, JSON schema, ...
- Variation contract test = test variations of the 200 OpenAPI specification (like different query params, headers, ...) + unhappy path (4xx, 5xx)  tests, for which you can define the targeted OpenAPI response.

For variation tests, you can instruct Portman which OpenAPI response you want to test against by setting the [openApiResponse](https://github.com/apideck-libraries/portman#variationtests-options)

> **openApiResponse (String | optional)** : References to the OpenAPI response object code/name for which a variation will be created. (example: "404"). If not defined, the 1st response object from OpenAPI will be taken as expected response. If the configured openApiResponse code is not defined in the OpenAPI document, Portman will not generate a variation for the targeted operations.

If the `openApiResponse` is defined, the "statusCode" test does not require the "code" parameter, since it will verify against the expected  `openApiResponse` code, but you can overwrite it if you want the tests to be generated based on the OpenAPI response but you expect your API to return a different HTTP code.

We have a variation example that showcases the usage `openApiResponse`, which hopefully makes it a bit clearer.
https://github.com/apideck-libraries/portman/tree/main/examples/testsuite-variation-tests#example-explained

**Reply by savage-alex** [reply]
Hi @thim81 thank you for the amazing reply as always. One thing I am unsure of from the above is the comment "the 200 request OpenAPI specification", The API definition I have used as input does not have a 200 but a 201

```post:
      operationId: CreateDesks
      summary: Create new desk
      description: Create a new desk not assigned to a specific user
      parameters:
        - $ref: 'https://api.swaggerhub.com/domains/AdvancedComputerSoft/advanced-common-oas3/1.3.0#/components/parameters/workplaceRefParam'
      tags:
        - Desks
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/CreateDeskDetail"
        required: true
      responses:
        "201":
          description: "Created"
          headers:
            Location:
              description: URL for newly created resource
              example: "https://myWorkplace.oneadvanced.com/desks/352d4069-2c17-4d82-9a5b-ba0ec43e3c9a"
              schema:
                type: string
```

Which resulted in a postman test of 

```
// Validate response status code 
pm.test("[GET]::/workplaces/:workplaceRef/desks - Response status code is 200", function () {
    pm.expect(pm.response.code).to.equal(200);
});
```

Is this expected?

**Reply by thim81** [reply]
@savage-alex this is good remark, I'll have a look at the code to verify the behaviour.

Is this a generated variation test? Because a normal contract test would result in a 2xx status check:
```js
// Validate status 2xx 
pm.test("[GET]::/lists/v2/audiences/:audienceId/contacts - Status code is 2xx", function () {
   pm.response.to.be.success;
});
```

Could you share with me the Portman configuration? 
So I can try to reproduce and potentially correct the behaviour.

**Reply by savage-alex** [reply]
Hi, its a core contract test only at the moment. Config attached (had to rename to txt to upload)
[portman-config.adv.simple.json.txt](https://github.com/apideck-libraries/portman/files/7869652/portman-config.adv.simple.json.txt)
d

**Reply by savage-alex** [reply]
let me know if you need any help with this

**Reply by thim81** [reply]
When running the normal contract tests with the below OpenAPI spec & Portman config, I get these tests:

![2022-01-14 at 12 33 38](https://user-images.githubusercontent.com/952446/149509350-ccae6cca-0308-47f4-b3a3-1580cabe0da8.png)

The contract tests are capable of handling a 201 (I'll see if I can optimise the 2xx & 201 check, since both are both overkill).
I'll now try to generate a variation and see what the result is.

```yaml
openapi: 3.0.3
info:
  version: "1"
  title: Workspace Desks
  contact: {}
paths:
  /workplaces/{workplaceRef}/desks:
    post:
      operationId: CreateDesks
      summary: Create new desk
      description: Create a new desk not assigned to a specific user
      parameters:
        - $ref: 'https://api.swaggerhub.com/domains/AdvancedComputerSoft/advanced-common-oas3/1.3.0#/components/parameters/workplaceRefParam'
      tags:
        - Desks
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/CreateDeskDetail"
        required: true
      responses:
        "201":
          description: "Created"
          headers:
            Location:
              description: URL for newly created resource
              example: "https://myWorkplace.oneadvanced.com/desks/352d4069-2c17-4d82-9a5b-ba0ec43e3c9a"
              schema:
                type: string
components:
  schemas:
    CreateDeskDetail:
      properties:
        desk:
          type: string
        values:
          additionalProperties:
            type: string
          type: object
      type: object
```

Portman config:
```json
{
  "version": 1.0,
  "tests": {
    "contractTests": [
      {
        "openApiOperation": "*::/*",
        "statusSuccess": {
          "enabled": true
        }
      },
      {
        "openApiOperation": "*::/*",
        "statusCode": {
          "enabled": true
        }
      },
      {
        "openApiOperation": "*::/*",
        "responseTime": {
          "enabled": true,
          "maxMs": 2000
        }
      },
      {
        "openApiOperation": "*::/*",
        "contentType": {
          "enabled": true
        }
      },
      {
        "openApiOperation": "*::/*",
        "jsonBody": {
          "enabled": true
        }
      },
      {
        "openApiOperation": "*::/*",
        "schemaValidation": {
          "enabled": true,
          "additionalProperties": false
        }
      },
      {
        "openApiOperation": "*::/*",
        "headersPresent": {
          "enabled": true
        }
      }
    ],
    "contentTests": [],
    "extendTests": []
  }
}
```

**Reply by savage-alex** [reply]
This is very interesting. When I use the same portman config I get Response code to equal 200 (hence the reason I raised this discussion)
![image](https://user-images.githubusercontent.com/25430683/149510690-b706b053-56c4-43f5-862f-aa0c29026a68.png)
are there any other configs that could infulence this?

**Reply by savage-alex** [reply]
let me try the openAPI you have shared. The difference between that and mine is that mine has 400x responses as well.

**Reply by savage-alex** [reply]
Ok. Its not to do with the 400x responses on this operation but actually when you add another operation on that path such as GET, it impacts the 201
[AdvancedComputerSoft-desks-post.yaml.txt](https://github.com/apideck-libraries/portman/files/7870170/AdvancedComputerSoft-desks-post.yaml.txt)
.

**Reply by thim81** [reply]
That seems like a bug. Using your YAML, I can reproduce it and I'll try to figure out where the bug happens.

**Reply by savage-alex** [reply]
Thanks @thim81, We are doing training on it so trying to make a super generic contract test we can throw at any API to get people started. Was a lightbulb moment today - can we check status codes as currently just doing the "statusSuccess" check as I had sadly seen OAI saying return a 201 and the API returning a 200 which it wouldnt currently be able to warn. 
Always a pleasure. Hope we can share a beer at APIDays or similar one day

**Reply by AliKhalili** [reply]
@thim81 Thank you for your amazing tool.
I tried to fix this issue here [241](https://github.com/apideck-libraries/portman/pull/241). Hope you find this useful.

**Reply by thim81** [reply]
The fix from @AliKhalili got merged in and will be part of the next release, which is planned in the coming days with some other small bug fixes.

**Comment by thim81** [comment]
FYI: In theory, you don't need the combination of `statusCode` + `statusSuccess` for a contract test, since the Postman assertion used: `pm.response.to.be.success;` will cover any HTTP 2xx (200,201,2..) response. 

By adding the `statusCode` check, you make the validation stricter, which can be desired if needed.

![2022-01-14 at 14 54 08](https://user-images.githubusercontent.com/952446/149526398-13d3e445-7fce-4966-b6e9-aa060d1d392d.png)

1) statusSuccess
2) statusCode


```json
{
  "version": 1.0,
  "tests": {
    "contractTests": [
      {
        "openApiOperation": "*::/*",
        "statusSuccess": {
          "enabled": true
        }
      },
      {
        "openApiOperation": "*::/*",
        "statusCode": {
          "enabled": true
        }
      },
```

**Reply by savage-alex** [reply]
100% we are intending to move from the success to the statusCode function for more stricter validation.

**Reply by thim81** [reply]
@savage-alex The statusCode fix from @AliKhalili has landed in the latest release 1.13.1 from Portman.

**Comment by thim81** [comment]
@savage-alex FYI: We inject contract tests (happy path) for all OpenAPI responses ranging from 200 till 302

So if you have a 201 or 204 or even a 302, this will be considered the happy path for the normal contract tests, if there are multiple "happy path" responses are defined, the 1ste happy response will be used as expected response for which Portman will generate the contract tests.

For variation tests, you can define the expected OpenAPI request using `openApiResponse` setting.

**Reply by savage-alex** [reply]
Answered my next question of what if you have more than one 200/300. It takes the first. Thank you

**Comment by savage-alex** [comment]
Tested this with the latest release and it works great. How can we close this dicussion please?

**Reply by thim81** [reply]
We can leave it here, for people we run into similar thoughts.
